File: books-server/src/main/java/com/jetbrains/bookstore/BookController.java
package com.jetbrains.bookstore;

import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/books")
class BookController {
    private final BookRepository bookRepository;

    BookController(BookRepository bookRepository) {
        this.bookRepository = bookRepository;
    }

    @GetMapping
    Books getAll() {
        Sort sort = Sort.by(Sort.Direction.DESC, "id");
        List<Book> books = bookRepository.findAll(sort);
        return Books.of(books);
    }

    @GetMapping("/{isbn}")
    ResponseEntity<Book> getByIsbn(@PathVariable String isbn) {
        return bookRepository.findByIsbn(isbn)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping(value = "/search", version = "1.0")
    Books searchBooks(@RequestParam("q") String query) {
        List<Book> books = bookRepository.findByTitleContainingIgnoreCase(query);
        return Books.of(books);
    }

    @GetMapping(value = "/search", version = "2.0")
    Books searchBooks_V2(@RequestParam("q") String query) {
        List<Book> books = bookRepository.searchBooks(query);
        return Books.of(books);
    }

    record Books(List<Book> books){
        public static Books of(List<Book> books) {
            return new Books(books);
        }
    }
}

File: books-server-kotlin/src/main/kotlin/com/jetbrains/demo/api/BookController.kt
/*
package com.jetbrains.demo.api

import com.jetbrains.demo.domain.Book
import com.jetbrains.demo.domain.BookRepository
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/books")
class BookController(private val bookRepository: BookRepository) {

    @GetMapping
    fun getAll(): Books {
        val books = bookRepository.findAll()
        return Books(books)
    }

    @GetMapping("/{isbn}")
    fun getByIsbn(@PathVariable isbn: String): ResponseEntity<Book> {
        return bookRepository.findByIsbn(isbn)
            .map{book -> ResponseEntity.ok(book)}
            .orElse(ResponseEntity.notFound().build())
    }

    @GetMapping(value = ["/search"], version = "1.0")
    fun searchBooks(@RequestParam("q") query: String): Books {
        val books = bookRepository.findByTitleContainingIgnoreCase(query)
        return Books(books)
    }

    @GetMapping(value = ["/search"], version = "2.0")
    fun searchBooksV2(@RequestParam("q") query: String): Books {
        val books = bookRepository.searchBooks("%$query%")
        return Books(books)
    }

    data class Books(val books: List<Book>)

}
*/
